keps/sig-network/3453-minimize-iptables-restore/README.md

```diff
@@ -0,0 +1,468 @@
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+There are no rollout/rollback-specific failure modes. If the code
+works, then it will work fine during rollout, and if the code is
+buggy, then it will (presumably) fail (to some extent) during rollout,
+presumably impacting already-running workloads.
+
+Given a sufficiently-catastrophic rollout failure, a rollback could
+fail because, eg, some component could no longer reach the apiserver
+because kube-proxy had written incorrect iptables rules. At this point
+there would be no easy way to recover the cluster. It is unlikely that
+such a drastic bug would make it to a GA release.
+
+Other than that, it should not be possible for a rollback to fail,
+because when the old/non-feature-enabled version of kube-proxy starts
+up, it would delete all of the rules created by the
+new/feature-enabled kube-proxy and replace them with the correct
+rules.
+
+###### What specific metrics should inform a rollback?
+
+We should add metrics to catch the failure modes noted above in [Risks
+and Mitigations](#risks-and-mitigations); in particular we should make
+it noticeable when failed partial resyncs occur, and when a partial
+resync would have generated a different result than a full resync.
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+<!--
+Describe manual testing that was done and the outcomes.
+Longer term, we may want to require automated upgrade/rollback tests, but we
+are missing a bunch of machinery and tooling and can't do that now.
+-->
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+No
+
+### Monitoring Requirements
+
+###### How can an operator determine if the feature is in use by workloads?
+
+This is a change to the core functioning of kube-proxy; if the feature
+is enabled, then it is in use.
+
+###### How can someone using this feature know that it is working for their instance?
+
+- [X] Other (treat as last resort)
+  - Details: End users do not "use this feature". It is a change to
+    the core functioning of kube-proxy. If the feature is enabled, and
+    their Services work, then the feature is working for them.
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+TBD
+
+sync performance should be much better in large clusters, and should
+be somewhat independent of cluster size
+
+`sync_proxy_rules_duration_seconds`
+`network_programming_duration_seconds`
+`sync_proxy_rules_endpoint_changes_pending`
+`sync_proxy_rules_service_changes_pending`
+`sync_proxy_rules_last_queued_timestamp_seconds` vs `sync_proxy_rules_last_timestamp_seconds`
+
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+- [X] Metrics
+  - Metric name: TBD
```

Comment:
Please D

Reply:
I think we can reuse current `sync_proxy_rules_iptables_restore_failures_total` metrics and label it to distinguis a partial_sync from a full_sync failure 

https://github.com/kubernetes/kubernetes/pull/110268/files#r967886374

Reply:
I commented on your comment on the other PR; it seems like that would be a metrics "API break" so I think having a new metric would be better.